<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
**Data Analyst Course Final Project - Student Project**
    
**E-Commerce Product Range Analysis**
    
**SQL Task**

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
**Task Description**
    
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.
    
    
**Description of Data**
    
books - contains data on books:
    
* book_id
    
* author_id
    
* title
    
* num_pages - number of pages
    
* publication_date
    
* publisher_id
    
    
authors - contains data on authors:
    
* author_id
    
* author
    

publishers - contains data on publishers:
    
* publisher_id
    
* publisher
    

ratings - contains data on user ratings:
    
* rating_id
    
* book_id
    
* username - the name of the user who rated the book
    
* rating
    
    
reviews - contains data on customer reviews:
    
* review_id
    
* book_id
    
* username - the name of the user who reviewed the book
    
* text - the text of the review

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
**Task**
    
* Find the number of books released after January 1, 2000.
    
* Find the number of user reviews and the average rating for each book.
    
* Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
    
* Identify the author with the highest average book rating (look only at books with at least 50 ratings).
    
* Find the average number of text reviews among users who rated more than 50 books. 
    

**Instructions for completing the task**
    
* Describe the goals of the study.
    
* Study the tables (print the first rows).
    
* Make an SQL query for each of the tasks.
    
* Output the results of each query in the Notebook.

* Describe your conclusions for each of the tasks.

In [19]:
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
First, we will load each table into its corresponding dataframe, and change column types if necessary.

In [20]:
query = '''
                select * from books
        '''

books_data = pd.io.sql.read_sql(query, con = engine)

books_data

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [21]:
books_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [22]:
books_data['publication_date'] = pd.to_datetime(books_data['publication_date'])
books_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   book_id           1000 non-null   int64         
 1   author_id         1000 non-null   int64         
 2   title             1000 non-null   object        
 3   num_pages         1000 non-null   int64         
 4   publication_date  1000 non-null   datetime64[ns]
 5   publisher_id      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


In [23]:
query = '''
                select * from authors
        '''

authors_data = pd.io.sql.read_sql(query, con = engine)

authors_data

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [24]:
query = '''
                select * from publishers
        '''

publishers_data = pd.io.sql.read_sql(query, con = engine)

publishers_data

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [25]:
query = '''
                select * from ratings
        '''

ratings_data = pd.io.sql.read_sql(query, con = engine)

ratings_data

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [26]:
query = '''
                select * from reviews
        '''

reviews_data = pd.io.sql.read_sql(query, con = engine)

reviews_data

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Now we will handle our first task - find the number of books released after January 1, 2000.

In [27]:
query_q1 = '''
SELECT COUNT(*) AS book_count
FROM books
WHERE publication_date > '2000-01-01'
'''

results_q1 = pd.io.sql.read_sql(query_q1, con = engine)
book_count = results_q1.at[0, 'book_count']

print('Number of books released after January 1, 2000:', book_count)

Number of books released after January 1, 2000: 819


<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Now we will handle our second task - find the number of user reviews and the average rating for each book.

In [28]:
merge_query_q2_updated = """
SELECT books.book_id, books.title, COALESCE(num_reviews, 0) AS num_reviews, COALESCE(avg_rating, 0) AS avg_rating
FROM books
LEFT JOIN (
    SELECT book_id, COUNT(review_id) AS num_reviews
    FROM reviews
    GROUP BY book_id
) AS subquery_reviews ON books.book_id = subquery_reviews.book_id
LEFT JOIN (
    SELECT book_id, AVG(rating) AS avg_rating
    FROM ratings
    GROUP BY book_id
) AS subquery_ratings ON books.book_id = subquery_ratings.book_id
"""
merged_df_q2_updated = pd.read_sql_query(merge_query_q2_updated, engine)
merged_df_q2_updated

book_id                                              title  num_reviews  \
0        652          The Body in the Library (Miss Marple  #3)            2   
1        273                                          Galápagos            2   
2         51                           A Tree Grows in Brooklyn            5   
3        951  Undaunted Courage: The Pioneering First Missio...            2   
4        839                                        The Prophet            4   
..       ...                                                ...          ...   
995       64                                Alice in Wonderland            4   
996       55          A Woman of Substance (Emma Harte Saga #1)            2   
997      148                                          Christine            3   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)            2   
999      828                           The Plot Against America            2   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Now we will handle our third task - identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [31]:
query_q3 = """
    SELECT publishers.publisher_id, publishers.publisher, COUNT(books.book_id) AS book_count
    FROM books
    JOIN publishers ON books.publisher_id = publishers.publisher_id
    WHERE books.num_pages > 50
    GROUP BY publishers.publisher_id, publishers.publisher
    ORDER BY book_count DESC
    LIMIT 1
"""
result_q3 = pd.read_sql_query(query_q3, engine)
result_q3

publisher_id      publisher  book_count
0           212  Penguin Books          42

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
We see that Penguin Books is our leading publisher for books with more than 50 pages, having 42 such books in our data.

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Now we will handle our fourth task - identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [32]:
query_q4 = """
    SELECT authors.author_id, authors.author, AVG(ratings.rating) AS avg_rating
    FROM books
    JOIN authors ON books.author_id = authors.author_id
    JOIN ratings ON books.book_id = ratings.book_id
    WHERE books.book_id IN (
        SELECT book_id
        FROM ratings
        GROUP BY book_id
        HAVING COUNT(*) >= 50
    )
    GROUP BY authors.author_id, authors.author
    ORDER BY avg_rating DESC
    LIMIT 1
"""
result_q4 = pd.read_sql_query(query_q4, engine)
result_q4

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
The author with the highest average book rating (at least 50 ratings) is the tandem of J.K. Rowling and Mary GrandPré, with an average rating of 4.28.

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Now we will handle our fifth task - find the average number of text reviews among users who rated more than 50 books.

In [33]:
query_q5 = """
    SELECT AVG(review_count) AS avg_text_reviews
    FROM (
        SELECT r.username, COUNT(DISTINCT rv.review_id) AS review_count
        FROM ratings r
        JOIN reviews rv ON r.username = rv.username
        WHERE r.username IN (
            SELECT username
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 50
        )
        GROUP BY r.username
    ) AS subquery
"""
result_q5 = pd.read_sql_query(query_q5, engine)
print("Average number of text reviews among users who rated more than 50 books:", result_q5['avg_text_reviews'].values[0])

Average number of text reviews among users who rated more than 50 books: 24.333333333333332


<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
Average number of text reviews among users who rated more than 50 books is 24.333333333333332.

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>
    
**Conclusions and Summary**
    
* We found out that 819 books were released after January 1, 2000.
    
* We created a joint table that has the number of reviews and avergae rating for each book.
    
* We found that Penguin Books is our leading publisher for books with more than 50 pages, having 42 such books in our data.
    
* We found that the author with the highest average book rating (at least 50 ratings) is the tandem of J.K. Rowling and Mary GrandPré, with an average rating of 4.28.
    
* We found that the average number of text reviews among users who rated more than 50 books is 24.333333333333332.